engine/src/main/scala/cromwell/engine/workflow/ShadowWorkflowManagerActor.scala

```diff
@@ -0,0 +1,255 @@
+  case object AbortAllWorkflowsCommand extends ShadowWorkflowManagerActorCommand
+  private[ShadowWorkflowManagerActor] final case class RestartWorkflowsCommand(workflows: Seq[WorkflowDescriptor]) extends ShadowWorkflowManagerActorCommand
+
+  /**
+    * Responses
+    */
+  sealed trait ShadowWorkflowManagerActorResponse extends ShadowWorkflowManagerActorMessage
+  // None yet?
+
+  def props(): Props = Props(new ShadowWorkflowManagerActor())
+
+  /**
+    * States
+    */
+  sealed trait ShadowWorkflowManagerState
+  case object Running extends ShadowWorkflowManagerState
+  case object Aborting extends ShadowWorkflowManagerState
+  case object Done extends ShadowWorkflowManagerState
+
+  /**
+    * Data
+    */
+  case class ShadowWorkflowManagerData(workflows: Map[WorkflowId, ActorRef]) {
+    def idFromActor(actor: ActorRef): Option[WorkflowId] =
+      workflows.collectFirst { case (id, a) if a == actor => id }
+    def withAddition(entry: WorkflowIdToActorRef): ShadowWorkflowManagerData =
+      this.copy(workflows = workflows + (entry.workflowId -> entry.workflowActor))
+    def without(id: WorkflowId): ShadowWorkflowManagerData =
+      this.copy(workflows = workflows - id)
+    def without(actor: ActorRef): ShadowWorkflowManagerData = {
+      // If the ID was found in the lookup return a modified copy of the state data, otherwise just return
+      // the same state data.
+      idFromActor(actor) map without getOrElse this
+    }
+  }
+
+  lazy val defaultConfig = ConfigFactory.load
+}
+
+/**
+  * Responses to messages:
+  * SubmitWorkflow: Returns a `Future[WorkflowId]`
+  *
+  */
+class ShadowWorkflowManagerActor(config: Config)
+  extends LoggingFSM[ShadowWorkflowManagerState, ShadowWorkflowManagerData] with CromwellActor {
+
+  def this() = this(ShadowWorkflowManagerActor.defaultConfig)
+  implicit val actorSystem = context.system
+
+  private val RestartDelay: FiniteDuration = 200 milliseconds
+  private val logger = Logging(context.system, this)
+  private val tag = self.path.name
+
+  private val donePromise = Promise[Unit]()
+
+  override def preStart() {
+    addShutdownHook()
+    restartIncompleteWorkflows()
+  }
+
+  private def addShutdownHook(): Unit = {
+    // Only abort jobs on SIGINT if the config explicitly sets backend.abortJobsOnTerminate = true.
+    val abortJobsOnTerminate =
+      config.getConfig("backend").getBooleanOr("abortJobsOnTerminate", default = false)
+
+    if (abortJobsOnTerminate) {
+      sys.addShutdownHook {
+        logger.info(s"$tag: Received shutdown signal. Aborting all running workflows...")
+        self ! AbortAllWorkflowsCommand
+        Await.ready(donePromise.future, Duration.Inf)
+      }
+    }
+  }
+
+  private def replyToAbort(id: WorkflowId): Unit = {
+    // Access the mutable state here synchronously, not in the callback.
+    val workflowActor = stateData.workflows.get(id)
+    val sndr = sender()
+    globalDataAccess.getWorkflowState(id) onComplete {
```

Comment:
One of the key abstractions of the actor model is that inside the actor only one thing is being processed at once. When you have futures (including callbacks) kicking off inside of an actor that no longer is the case, you're back to needing to reason about things like shared mutable state and other concurrency issues - there are better schemes for that than actors.

It's not the end of the world and sometimes it is unavoidable (e.g. if you look at things like pipeTo, they're also using callbacks albeit in a more controlled fashion) but the fewer places it is used the fewer places one needs to reason about that class of concurrency issue.

An example like this one would appear to be in the 'no big deal' category but the problem is that as we've seen it's easy for stuff to slowly creep in